### PR TITLE
Cross validation

### PR DESCRIPTION
**Pull request by VimWizardVersace**
Add a new field to perform cross validation across the TRAIN partition of data via the n_cv_folds param. 

This PR also resets the index of the data post filtering. We filter out certain rows with only one sample per tree as we can not calculate variance from one sample, but previously these just left empty gaps in the index. Reindexing was necessary for cross validation.

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -166,6 +168,64 @@ class SoftplusLayer(tf.keras.layers.Layer):
     def call(self, inputs):
         return tf.math.log(1 + tf.exp(inputs))
 
+def cross_val_with_best_model(
+    model_fn: Callable,
```
Comment (rothn) [inline model.py:172]: Could you specify more concretely what this needs to return? e.g. `Callable[SomeAbstractClass]`.

File: model.py [hunk 2]
```diff
@@ -166,6 +168,64 @@ class SoftplusLayer(tf.keras.layers.Layer):
...
+    sp: ScaledPartitions,
+    n_cv_folds: int,
+    epochs_per_fold: int,
+    batch_size: int,
+    callbacks_list: List[Callable]):
+    '''
+        Runs cross validation on the model, returning the history and fitted best model. 
+
+        real: Function that builds, compiles, and returns a model.
+        sp: Stands for "Scaled Partitions", representing your dataset. This function only looks at the "train"
```
Comment (rothn) [inline model.py:182]: Is the intended use-case here that most folks using cross-validation will use a `ScaledPartitions` object with only a training set?

File: model.py [hunk 3]
```diff
@@ -166,6 +168,64 @@ class SoftplusLayer(tf.keras.layers.Layer):
     def call(self, inputs):
         return tf.math.log(1 + tf.exp(inputs))
 
+def cross_val_with_best_model(
+    model_fn: Callable,
+    sp: ScaledPartitions,
+    n_cv_folds: int,
+    epochs_per_fold: int,
+    batch_size: int,
```
Comment (rothn) [inline model.py:176]: Perhaps we could add a `fit_kwargs` field here so that we don't need DNN/Keras specifics like `batch_size` and `callbacks_list`, but still can support direct training on these APIs without adding a wrapper class.

Then, we would invoke `model.fit(X_train, Y_train, validation_data=(X_val, Y_val,), epochs=epochs_per_fold, **fit_kwargs)`

File: dataset.py [hunk 4]
```diff
@@ -560,9 +561,11 @@ def load_and_scale(config: Dict,
                    extra_columns_from_geotiffs: Dict[str, raster.AmazonGeoTiff]) -> ScaledPartitions:
   columns_to_keep = columns_to_passthrough + columns_to_scale + columns_to_standardize
   X_train, Y_train = load_dataset(config['TRAIN'], mean_label, var_label, columns_to_keep, extra_columns_from_geotiffs)
-  X_val, Y_val = load_dataset(config['VALIDATION'], mean_label, var_label, columns_to_keep, extra_columns_from_geotiffs)
   X_test, Y_test = load_dataset(config['TEST'], mean_label, var_label, columns_to_keep, extra_columns_from_geotiffs)
-
+  
+  # Optionally load validation dataset. 
+  X_val, Y_val = None, None if 'VALIDATION' not in config else load_dataset(config['VALIDATION'], mean_label, var_label, columns_to_keep, extra_columns_from_geotiffs)
+
```
Comment (rothn) [inline dataset.py:568]: When using cross-validation, would it also make sense for the test set to be optional?

**Review by rothn — CHANGES_REQUESTED** [review]

File: model.py (hunk 2 above)
Comment (VimWizardVersace) [reply model.py:206]: That's right. We use it as a container mapping scaling values to the dataset.

File: model.py [hunk 5]
```diff
@@ -177,52 +254,76 @@ def train_or_update_variational_model(
...
+                model_file,
+                custom_objects={"KLCustomLoss": KLCustomLoss})
+        model.save(model_file)
+        dump(sp.feature_scaler, f"{model_file.strip('.keras')}.pkl")
+        packaged_model = TFModel(model_file, f"{model_file.strip('.keras')}.pkl")
+        return packaged_model
+    
+    if sp.val: 
+        # No cross validation
+        assert not n_cv_folds, "Cross validation not possible if manually specifying a validation dataset."
```
Comment (rothn) [inline model.py:309]: Prefer to raise an exception-- asserts are only guaranteed to be evaluated in debug mode.

File: model.py [hunk 6]
```diff
@@ -177,52 +254,76 @@ def train_or_update_variational_model(
...
+        model = build_model()
+        history = model.fit(sp.train.X, sp.train.Y, verbose=0, validation_data=sp.val.as_tuple(), shuffle=True,
+                            epochs=epochs, batch_size=batch_size, callbacks=callbacks_list)
+        return history, model
+    else:
+        # Yes cross validation
+        assert n_cv_folds, "If no validation dataset is specified, number of cross validation folds must be set."
```
Comment (rothn) [inline model.py:316]: Prefer to raise an exception-- asserts are only guaranteed to be evaluated in debug mode.

File: model.py [hunk 7]
```diff
@@ -166,6 +191,58 @@ class SoftplusLayer(tf.keras.layers.Layer):
...
+
+        if score > best_score:
+            best_score = score
+            best_model_artifacts = training_artifacts
+            if cache_best_model:
+                best_model = deepcopy(model)
+            if output_model_filename:
+                model.save_model(output_model_filename) 
+
+    return training_artifacts, best_model
```
Comment (rothn) [inline model.py:243]: It's perfectly valid (and will produce better results) to return a model trained on the entire train+validation set after performing cross-validation. Could we do this here?

**Review by rothn — CHANGES_REQUESTED** [review]
Thanks so much! Just a few minor things now!

**Review by rothn — APPROVED** [review]
Looks great! Thanks for sticking with this!
